vir/simd_permute.h

```diff
@@ -0,0 +1,229 @@
+
+#include "simd.h"
+#include <bit>
+
+namespace vir
+{
+  namespace detail
+  {
+    template <typename F>
+      concept index_permutation_function_nosize = requires(F const& f)
+      {
+	{ f(std::integral_constant<std::size_t, 0>()) } -> std::integral;
+      };
+
+    template <typename F, std::size_t Size>
+      concept index_permutation_function_size = requires(F const& f)
+      {
+	{ f(std::integral_constant<std::size_t, 0>(), std::integral_constant<std::size_t, Size>()) }
+	  -> std::integral;
+      };
+
+    template <typename F, std::size_t Size>
+      concept index_permutation_function
+	= index_permutation_function_size<F, Size> or index_permutation_function_nosize<F>;
+  }
+
+  constexpr int simd_permute_zero = INT_MAX;
+
+  constexpr int simd_permute_uninit = INT_MAX - 1;
+
+  namespace simd_permutations
+  {
+    struct DuplicateEven
+    {
+      consteval unsigned
+      operator()(unsigned i) const
+      { return i & ~1u; }
+    };
+
+    inline constexpr DuplicateEven duplicate_even {};
+
+    struct DuplicateOdd
+    {
+      consteval unsigned
+      operator()(unsigned i) const
+      { return i | 1u; }
+    };
+
+    inline constexpr DuplicateOdd duplicate_odd {};
+
+    template <unsigned N>
+      struct SwapNeighbors
+      {
+	consteval unsigned
+	operator()(unsigned i, auto size) const
+	{
+	  static_assert(size % (2 * N) == 0,
+			"swap_neighbors<N> permutation requires a multiple of 2N elements");
+	  if (std::has_single_bit(N))
+	    return i ^ N;
+	  else if (i % (2 * N) > N)
+	    return i - N;
+	  else
+	    return i + N;
+	}
+      };
+
+    template <unsigned N = 1u>
+      inline constexpr SwapNeighbors<N> swap_neighbors {};
+
+    template <int Position>
+      struct Broadcast
+      {
+	consteval int
+	operator()(int) const
+	{ return Position; }
+      };
+
+    template <int Position>
+      inline constexpr Broadcast<Position> broadcast {};
+
+    inline constexpr Broadcast<0> broadcast_first {};
+
+    inline constexpr Broadcast<-1> broadcast_last {};
+
+    struct Reverse
+    {
+      consteval int
+      operator()(int i) const
+      { return -1 - i; }
+    };
+
+    inline constexpr Reverse reverse {};
+
+    template <int Offset>
+      struct Rotate
+      {
+	consteval int
+	operator()(int i, auto size) const
+	{ return (i + Offset) % size(); }
+      };
+
+    template <int Offset>
+      inline constexpr Rotate<Offset> rotate {};
+
+    template <int Offset>
+      struct Shift
+      {
+	consteval int
+	operator()(int i, int size) const
+	{
+	  const int j = i + Offset;
+	  if constexpr (Offset >= 0)
+	    return j >= size ? simd_permute_zero : j;
+	  else
+	    return j < 0 ? simd_permute_zero : j;
+	}
+      };
+
+    template <int Offset>
+      inline constexpr Shift<Offset> shift {};
+  }
+
+  template <std::size_t N = 0, vir::any_simd_or_mask V,
+	    detail::index_permutation_function<V::size()> F>
+    VIR_ALWAYS_INLINE constexpr stdx::resize_simd_t<N == 0 ? V::size() : N, V>
```

Comment:
Why "always" inline? Is there a likely case where the compiler would not inline this at all? Or do you want to avoid symbols in the final object that might depend on compiler flags and conflict with other symbols of the same name?

Anyway: Maybe explain this somewhere?

Reply:
debug builds. Consider `simd` to be the missing fundamental type. You don't get function calls for builtin operations on `int` either. There's also constant propagation to consider here. A combination of operations involving one (or more) permute(s) might be optimized into something very different by the compiler if it considers everything together. Also I've been bitten recently: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=108030. If the compiler gets it wrong you can pay in factors of performance, not percentages.